### PR TITLE
Quotas management

### PR DESCRIPTION
**Pull request by firepolo**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Fixed issues | comma-separated list of issues fixed by the PR, if any

# Description
Add a feature for add quota managment in Claroline, we can activate or not the option.

# Screenshots
Add tool in training menu
image

Add options in session parameters
image

Add view for list & edit quotas
image

Add view for list & manage subscriptions
image
image

### CODE REVIEW

File: src/plugin/cursus/Resources/modules/tools/trainings/quota/store/actions.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+actions.openForm = (id = null, defaultProps = {}) => (dispatch) => {
+  if (!id) {
+    return dispatch(formActions.resetForm(selectors.FORM_NAME, defaultProps, true))
+  }
+
+  return dispatch({
+    [API_REQUEST]: {
+      url: url(['apiv2_cursus_quota_find'], {filters: {uuid: id}}),
```
Comment (Elorfin) [inline src/plugin/cursus/Resources/modules/tools/trainings/quota/store/actions.js:36]: use `get` endpoint instead.

File: src/plugin/cursus/Resources/modules/tools/trainings/quota/store/actions.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+export const actions = {}
+
+actions.loadQuota = makeActionCreator(LOAD_QUOTA, 'quota')
+
+actions.open = (id, force = false) => (dispatch, getState) => {
+  const currentQuota = selectors.quota(getState())
+  if (force || isEmpty(currentQuota) || currentQuota.id !== id) {
+    return dispatch({
+      [API_REQUEST]: {
+        url: ['apiv2_cursus_quota_open', {id}],
```
Comment (Elorfin) [inline src/plugin/cursus/Resources/modules/tools/trainings/quota/store/actions.js:20]: use `get` endpoint instead.

File: src/plugin/cursus/Resources/modules/tools/trainings/subscription/components/page.jsx [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  useEffect(() => {
+    if (!isEmpty(props.quota)) {
+      props.getStatistics(props.quota.id)
+    }
+  }, [props.quota])
+
+  return (
+    <PageFull
```
Comment (Elorfin) [inline src/plugin/cursus/Resources/modules/tools/trainings/subscription/components/page.jsx:34]: Same comment than for Quota

File: src/plugin/cursus/Resources/modules/tools/trainings/subscription/containers/main.jsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import {connect} from 'react-redux'
+import {withReducer} from '#/main/app/store/reducer'
+import {selectors, actions, reducer} from '#/plugin/cursus/tools/trainings/quota/store'
```
Comment (Elorfin) [inline src/plugin/cursus/Resources/modules/tools/trainings/subscription/containers/main.jsx:3]: You shouldn't request a part of another sub app.

**Review by Elorfin — CHANGES_REQUESTED** [review]
I still need to test it and check why status management is so complex.

File: src/main/core/Resources/modules/data/types/organization/components/input.jsx [hunk 5]
```diff
@@ -17,7 +17,7 @@ const OrganizationButton = props =>
     style={{marginTop: 10}}
     type={MODAL_BUTTON}
     icon="fa fa-fw fa-plus"
-    label={trans('add_organization')}
+    label={trans('set_organization')}
```
[the remark this answers is not in the document]
Comment (firepolo) [reply src/main/core/Resources/modules/data/types/organization/components/input.jsx:20]: I fix

File: src/main/core/API/Serializer/Platform/ClientSerializer.php [hunk 6]
```diff
@@ -141,6 +141,7 @@ public function serialize()
             ],
             'privacy' => $this->config->getParameter('privacy'),
             'pricing' => $this->config->getParameter('pricing'),
+            'quotas' => $this->config->getParameter('quotas'),
```
[the remark this answers is not in the document]
Comment (firepolo) [reply src/main/core/API/Serializer/Platform/ClientSerializer.php:144]: I remove the line

File: src/plugin/cursus/Controller/QuotaController.php [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+    public function getIgnore()
+    {
+        return ['copyBulk', 'doc', 'exist'];
+    }
+
+    /**
+     * @Route("/", name="apiv2_cursus_quota_list", methods={"GET"})
+     */
+    public function listAction(Request $request, $class = Quota::class): JsonResponse
```
[the remark this answers is not in the document]
Comment (firepolo) [reply src/plugin/cursus/Controller/QuotaController.php:88]: I remove this function

File: src/plugin/cursus/Entity/Session.php [hunk 8]
```diff
@@ -112,6 +122,11 @@ public function __construct()
         $this->events = new ArrayCollection();
     }
 
+    /**
+     * @ORM\Column(name="used_by_quotas", type="boolean")
+     */
+    private $usedByQuotas = false;
```
[the remark this answers is not in the document]
Comment (firepolo) [reply src/plugin/cursus/Entity/Session.php:123]: Yes, I you agree. I have no time for this refactor now. But I can do this later.

File: src/plugin/cursus/Repository/SessionUserRepository.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+namespace Claroline\CursusBundle\Repository;
+
+use Claroline\CursusBundle\Entity\Quota;
+use Doctrine\ORM\EntityRepository;
+
+class SessionUserRepository extends EntityRepository
+{
+    public function findByQuota(Quota $quota)
```
[the remark this answers is not in the document]
Comment (firepolo) [reply src/plugin/cursus/Repository/SessionUserRepository.php:19]: I renamed this

File: src/plugin/cursus/Resources/modules/quota/components/list.jsx [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    primaryAction={(row) => ({
+      type: LINK_BUTTON,
+      label: trans('open', {}, 'actions'),
+      target: `${props.path}/${row.id}`
+    })}
+    definition={[
+      {
+        name: 'organization.name',
+        alias: 'organization',
+        type: 'string',
```
[the remark this answers is not in the document]
Comment (firepolo) [reply src/plugin/cursus/Resources/modules/quota/components/list.jsx:30]: It's done, but I have error with the property at "primary: true". Can I keep it primary and have the primaryAction ?

**Comment by Elorfin** [comment]
This contains to many changes to be merge.
If you want to add it to the code base, you'll need to break it in multiple PRs.

Also note ;
- We will not accept new dependencies. All the export of Claroline are done in CSV which is natively supported by MSExcel,
- No extensions for Doctrine.
- No new endpoints in the core API.

**Comment by maieutiquer** [comment]
Temporarily restoring the branch in order to check if we have everything on our side as well

**Pull request closed without merging**